qubesusbproxy/core3ext.py

```diff
@@ -41,36 +44,192 @@
-                # don't cache this value
-                return "Unknown - domain not running"
-            untrusted_device_desc = self.backend_domain.untrusted_qdb.read(
+    def manufacturer(self) -> str:
+        """
+        The name of the manufacturer of the device introduced by device itself
+
+        Could be empty string or "unknown".
+
+        Lazy loaded.
+        """
+        if self._manufacturer is None:
+            result = self._load_desc_from_qubesdb()["manufacturer"]
+        else:
+            result = self._manufacturer
+        return result
+
+    @property
+    def name(self) -> str:
+        """
+        The name of the device it introduced itself with (could be empty string)
+
+        Could be empty string or "unknown".
+
+        Lazy loaded.
+        """
+        if self._name is None:
+            result = self._load_desc_from_qubesdb()["name"]
+        else:
+            result = self._name
+        return result
+
+    @property
+    def serial(self) -> str:
+        """
+        The serial number of the device it introduced itself with.
+
+        Could be empty string or "unknown".
+
+        Lazy loaded.
+        """
+        if self._serial is None:
+            result = self._load_desc_from_qubesdb()["serial"]
+        else:
+            result = self._serial
+        return result
+
+    @property
+    def interfaces(self) -> List[qubes.devices.DeviceInterface]:
+        """
+        List of device interfaces.
+
+        Every device should have at least one interface.
+        """
+        if self._interfaces is None:
+            result = self._load_interfaces_from_qubesdb()
+        else:
+            result = self._interfaces
+        return result
+
+    @property
+    def parent_device(self) -> Optional[qubes.devices.DeviceInfo]:
+        """
+        The parent device if any.
+
+        USB device has no parents.
+        """
+        return None
+
+    def _load_interfaces_from_qubesdb(self) \
+            -> List[qubes.devices.DeviceInterface]:
+        result = [qubes.devices.DeviceInterface.unknown()]
+        if not self.backend_domain.is_running():
+            # don't cache this value
+            return result
+        untrusted_interfaces: bytes = (
+            self.backend_domain.untrusted_qdb.read(
+                self._qdb_path + '/interfaces')
+        )
+        if not untrusted_interfaces:
+            return result
+        self._interfaces = result = [
+            qubes.devices.DeviceInterface(
+                self._sanitize(ifc, safe_chars=string.hexdigits), devclass="usb"
+            )
+            for ifc in untrusted_interfaces.split(b':')
+            if ifc
+        ]
+        return result
+
+    def _load_desc_from_qubesdb(self) -> Dict[str, str]:
+        unknown = "unknown"
+        result = {"vendor": unknown,
+                  "product": unknown,
+                  "manufacturer": unknown,
+                  "name": unknown,
+                  "serial": unknown}
```

Comment:
I noticed that this does _not_ include the port the device is attached through.  Qubes Manager will need this to display a GUI asking the user what device they plugged into a port.

Reply:
This is only data provided by the device itself (possibly fake/malicious), the port number is contained within 'ident' and is independent of the device itself (the device does not inform us which port it is connected to).